doc/content/xapi-guard/_index.md

```diff
@@ -26,3 +28,57 @@ Xapi-guard forwards calls from domains to xapi to persist UEFI variables, and up
 To do this, it listens to 1 socket per service (varstored, or swtpm) per domain.
 To create these sockets before the domains are running, it listens to a message-switch socket.
 This socket listens to calls from xenopsd, which orchestrates the domain creation.
+
+To protect the domains from xapi being unavailable transiently, xapi-guard provides an on-disk cache for vTPM writes.
+This cache acts as a buffer and stores the requests temporarily until xapi can be contacted again.
+This situation usually happens when xapi is being restarted as part of an update.
+SWTPM, the vTPM daemon, reads the contents of the TPM from xapi-guard on startup, suspend, and resume.
+During normal operation SWTPM does not send read requests from xapi-guard.
```

Comment:
It would be good if we added some assertions to check that this is really the case. It may match observed behaviour (i.e. swtpm has a cache of its own), and from a quick look at the source code it always tries to use the local cache, and only if that is empty it falls back to reading from the backend.

However this is only true for each TPM state individually (out of the 3), so you could still see mixed writes and reads between different TPM states.

Reply:
See https://github.com/xapi-project/xen-api/pull/5460/files#r1504263947 I think xapi-guard will create this scenario itself, it always uses read-modify-write and can lead to lost states if more than 1 of the 3 vTPM states is updated while xapi is down!
So we need to handle the case where we've got mixed reads and writes (and add a test for this scenario).